common/src/proof.rs

```diff
@@ -113,34 +117,77 @@ impl StarkProof {
 // ================================================================================================
 
 impl Queries {
-    /// Returns a set of queries constructed from a batch Merkle proof and corresponding values.
-    pub fn new<E: FieldElement>(merkle_proof: BatchMerkleProof, values: Vec<Vec<E>>) -> Self {
+    /// Returns a set of queries constructed from a batch Merkle proof and corresponding elements.
+    pub fn new<E: FieldElement>(merkle_proof: BatchMerkleProof, query_values: Vec<Vec<E>>) -> Self {
+        assert!(!query_values.is_empty(), "query values cannot be empty");
+        let elements_per_query = query_values[0].len();
+        assert!(
+            elements_per_query != 0,
+            "a query must contain at least one value"
+        );
+
         // TODO: add debug check that values actually hash into the leaf nodes of the batch proof
-        Queries {
-            paths: merkle_proof.nodes,
-            values: values
-                .into_iter()
-                .map(|v| E::elements_as_bytes(&v).to_vec())
-                .collect(),
+
+        // concatenate all elements together into a single vector of bytes
+        let num_queries = query_values.len();
+        let mut values = Vec::with_capacity(num_queries * elements_per_query * E::ELEMENT_BYTES);
+        for elements in query_values.iter() {
+            assert_eq!(
+                elements.len(),
+                elements_per_query,
+                "all queries must contain the same number of values"
+            );
+            values.extend_from_slice(E::elements_as_bytes(elements));
         }
+
+        // serialize internal nodes of the batch Merkle proof; we care about internal nodes only
+        // because leaf nodes can be reconstructed from hashes of query values
+        let paths = merkle_proof.serialize_nodes();
+
+        Queries { paths, values }
     }
 
-    /// Convert a set of queries into a batch Merkle proof and corresponding values.
-    /// TODO: return values as a vector of field elements
-    pub fn into_batch<H: Hasher>(self, num_leaves: usize) -> (BatchMerkleProof, Vec<Vec<u8>>) {
+    /// Convert a set of queries into a batch Merkle proof and corresponding query values.
+    pub fn deserialize<H: Hasher, E: FieldElement>(
```

Comment:
Can any zero value or length cause issues or panic (ie division by zero?)

Reply:
Good point! I've added some assertions here as this code gets called internally from within Winterfell. So, if it is called with wrong parameters, it's ok to panic.